system/Database/BaseBuilder.php

```diff
@@ -2051,6 +2067,72 @@ public function onConstraint($set)
         return $this;
     }
 
+    /**
+     * Sets data source as a query for insert/update/upsert
+     *
+     * @param BaseBuilder|RawSql|string $query
+     */
+    public function fromQuery($query): BaseBuilder
+    {
+        if (! empty($query)) {
+            if ($query instanceof BaseBuilder) {
+                $query = $query->getCompiledSelect();
+            } elseif ($query instanceof RawSql) {
+                $query = $query->__toString();
+            }
+
+            if (is_string($query)) {
+                $this->QBOptions['fromQuery'] = $query;
+                $this->QBKeys                 = $this->db->protectIdentifiers($this->fieldsFromQuery($query));
+                $this->QBSet                  = [];
+            }
+        }
+
+        return $this;
+    }
+
+    /**
+     * Gets column names from a select query
+     */
+    protected function fieldsFromQuery(string $sql): array
+    {
+        $sql = preg_replace('/\\(([^()]*+|(?R))*\\)/', '', $sql); // remove everything in parenthesis - removes "FROM" and commas
+        $d   = ['`', "'", '"'];
+
+        $o   = ' ';
```

Comment:
What is o?

Reply:
I can't remember what o was suppose to stand for..

Reply:
I remember.. o is for original character